src/tests/client.ts

```diff
@@ -329,6 +329,51 @@ it('should not send the complete message if the socket is not open', async () =>
   await sub.waitForComplete();
 });
 
+it('should use a custom json reviver and replacer function', async () => {
```

Comment:
You are actually not testing the `replacer` function, only the `reviver`. You can do so by using the `server.waitForMessage` and expecting the replacer results there.

Reply:
Ah my mistake, I thought it would be implicitly tested by transforming the keys of the JSON messages like `type` => `TYPE`. Just tested though and you are right, the replacer only replaces values:

```
> const uppercaser = (_, value) => {
...       if (typeof value == 'string') {
.....         return value.toUpperCase();
.....       }
...       return value;
...     }
undefined
> JSON.stringify({foo: "bar"}, uppercaser)
'{"foo":"BAR"}'
```

Will fix

Reply:
I prefer tests being isolated to the problem or scope. For example, if only the server starts failing, due to this integration, the client tests will fail in addition to the server's (2 failed tests, 1 scope). You wouldnt know whats really failing until you poke around. However, if the tests were kept isolated, the client will fail independently of the server and vice versa (1 failed test, 1 scope; 2 failed tests, 2 scopes). Having said that, please test the client here exclusively. (I dont mind the single test for both `reviver` and `replacer`, just dont mix the server in)

You can always transform a different message to avoid validation issues (like the `query { value }` when `MessageType = subscription`). Yeah, might be a few extra lines of code, but we keep the tests scoped.

Reply:
Oh and,
> That said, these options are still necessary on the server because the client will uppercase the connection init and downcase the connection ack message as well, so the server needs to properly transform those in order to establish the connection that then lets the actual test message work.

Keeping it simple, you can just separate the `replacer` test, check the message with `server.waitForMessage`, and let it fail due to validation. Still focused, and no harm done.

Reply:
I can't find an implementation of `server.waitForMessage` -- am I blind or are you maybe thinking of `client.waitForMessage`? I tried writing my own `server.waitForMessage`, but it's tricky because the first thing the server does when it goes to handle a message is try to parse it, which validates it. If there's no server side reviver that undoes whatever the client does to the message in the test then this validation fails and no event ever gets emitted for the test to pick up, error or otherwise. I haven't come up with a good replacer that makes a noticeable difference in the JSON that the test can assert on that doesn't also break the expectations of the library if the changes aren't undone before validation. 

That's why I wrote the less-than-ideally-isolated test, it just seems a bit easier to let the system work as intended to test this otherwise not super important feature along the happy path, since really it's the JSON builtin that is doing the actual work and we just need to test that the message was transformed at all.

Reply:
My bad! Had `startTServer` in my head. 😅

Please see https://github.com/enisdenjo/graphql-ws/pull/172#issuecomment-829520428.